### ISSUE TITLE
Cannot automatically set CgroupDriver when starting the Kubelet: cannot execute 'docker info -f {{.CgroupDriver}}': executable file not found in $PATH

### ISSUE THREAD

**Issue by imoisharma** [kubernetes/kubernetes#99053]
#### What happened:
Cgroup driver is not detected by Kubelet on Control Plane Node 
![image](https://user-images.githubusercontent.com/30807347/107839546-bd4e1b00-6e00-11eb-958e-214e3588076c.png)

#### What you expected to happen:
Containerd Driver should be detected by Kubelet

#### Anything else we need to know?:
I edit the config.yaml file which is present under the Kubelet. And did the same step mentioned but it's not working
```
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
cgroupDriver: <value>
```

https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/

#### Environment:
- Kubernetes version (use `kubectl version`): 1.20.1
- Cloud provider or hardware configuration:  AWS
- OS (e.g: `cat /etc/os-release`): 
NAME="Ubuntu"
VERSION="20.04.1 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.1 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

- Kernel (e.g. `uname -a`):  Linux ip-172-31-93-56 5.4.0-1029-aws #30-Ubuntu SMP Tue Oct 20 10:06:38 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

- Install tools: container, calico etc
- Network plugin and version (if this is a network-related bug): calico network 
- Others:
Steps I performed

cat <<EOF | sudo tee /etc/modules-load.d/containerd.conf
overlay
br_netfilter
EOF

sudo modprobe overlay

sudo modprobe br_netfilter

cat <<EOF | sudo tee /etc/sysctl.d/99-kubernetes-cri.conf
net.bridge.bridge-nf-call-iptables = 1
net.ipv4.ip_forward = 1
net.bridge.bridge-nf-call-ip6tables = 1
EOF

sudo sysctl --system

### Install and configure containerd

sudo apt-get update && sudo apt-get install -y containerd<br>
sudo mkdir -p /etc/containerd<br>
sudo containerd config default | sudo tee /etc/containerd/config.toml<br>
sudo systemctl restart containerd<br>

### On all nodes, disable swap

sudo swapoff -a<br>
```sudo sed -i '/ swap / s/^\(.*\)$/#\1/g' /etc/fstab```

### On all nodes, install kubeadm, kubelet, and kubectl<br>
sudo apt-get update && sudo apt-get install -y apt-transport-https curl<br>
curl -s https://packages.cloud.google.com/apt/doc/apt-key.gpg | sudo apt-key add -<br>

cat <<EOF | sudo tee /etc/apt/sources.list.d/kubernetes.list<br>
deb https://apt.kubernetes.io/ kubernetes-xenial main<br>
EOF<br>

sudo apt-get update<br>
sudo apt-get install -y kubelet=1.20.1-00 kubeadm=1.20.1-00 kubectl=1.20.1-00<br>
sudo apt-mark hold kubelet kubeadm kubectl<br>

### On the control plane node only, initialize the cluster and set up kubectl access.

sudo kubeadm init --pod-network-cidr 192.168.0.0/16<br>
mkdir -p $HOME/.kube<br>
sudo cp -i /etc/kubernetes/admin.conf $HOME/.kube/config<br>
sudo chown $(id -u):$(id -g) $HOME/.kube/config<br>

### Verify the cluster is working
kubectl version<br>

### Install the Calico network add-on.
kubectl apply -f https://docs.projectcalico.org/manifests/calico.yaml<br>

### Check the calico-related kube-system Pods to verify that everything is working so far (they may take a few moments to fully start up).
kubectl get pods -n kube-system

### Get the join command (this command is also printed during kubeadm init . Feel free to simply copy it from there).
kubeadm token create --print-join-command

/sig needs-sig

**Comment by BenTheElder** [comment]
/sig cluster-lifecycle
> I edit the config.yaml file which is present under the Kubelet. And did the same step mentioned but it's not working

Ah the docs here are a bit confusing:
https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#configure-cgroup-driver-used-by-kubelet-on-control-plane-node

You need to pass those values to the `--config` flag of `kubeadm init` like:

- create config file `kubeadm-init-config.yaml`
- add this kubelet snippet to it
  - if you need to do other kubeadm config objects, you can seperate them with a `---` line (yaml document seperator)
- `kubeadm init --config=kubeadm-init-config.yaml`

**NOT** editing the kubelet file before or after calling kubeadm, though they have the same format, and the input to `kubeadm init` will control how the final config is generated.

> Containerd Driver should be detected by Kubelet

This is also confusing -- _kubelet_ does not detect this, kubeadm respects the config. `kubeadm` detects this, and the log line you're seeing is from kubeadm despite being a file "`kubelet.go`", this comes from kubeadm here:
https://github.com/kubernetes/kubernetes/blob/f2be0bc1c719b0496ab22e54a319117e272cbd82/cmd/kubeadm/app/componentconfigs/kubelet.go#L200

**Comment by imoisharma** [comment]
Pardon me, but which snippet you want me to add because I added the snippet from the doc but it's not working. 
![image](https://user-images.githubusercontent.com/30807347/107845537-5b0d0e80-6e30-11eb-9b53-9bd2c1e52e8b.png)
![image](https://user-images.githubusercontent.com/30807347/107845549-66f8d080-6e30-11eb-817a-78959e223cf0.png)
would you please correct me what's wrong steps I am performing?

**Comment by imoisharma** [comment]
> /sig cluster-lifecycle
> 
> > I edit the config.yaml file which is present under the Kubelet. And did the same step mentioned but it's not working
> 
> Ah the docs here are a bit confusing:
> https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#configure-cgroup-driver-used-by-kubelet-on-control-plane-node
> 
> You need to pass those values to the `--config` flag of `kubeadm init` like:
> 
>     * create config file `kubeadm-init-config.yaml`
> 
>     * add this kubelet snippet to it
>       
>       * if you need to do other kubeadm config objects, you can seperate them with a `---` line (yaml document seperator)
> 
>     * `kubeadm init --config=kubeadm-init-config.yaml`
> 
> 
> **NOT** editing the kubelet file before or after calling kubeadm, though they have the same format, and the input to `kubeadm init` will control how the final config is generated.
> 
> > Containerd Driver should be detected by Kubelet
> 
> This is also confusing -- _kubelet_ does not detect this, kubeadm respects the config. `kubeadm` detects this, and the log line you're seeing is from kubeadm despite being a file "`kubelet.go`", this comes from kubeadm here:
> 
> https://github.com/kubernetes/kubernetes/blob/f2be0bc1c719b0496ab22e54a319117e272cbd82/cmd/kubeadm/app/componentconfigs/kubelet.go#L200

check above

**Comment by BenTheElder** [comment]
the config file should contain only:
```
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
cgroupDriver: <value>
```

**NOTE**: replace `<value>` with your actual value, valid values are `systemd` and `cgroupfs`. The default is `cgroupfs`

your `config.yaml` contains many other fields that should be generated by kubeadm instead.

... but you also only need to do this if containerd is configured to use the systemd cgroup driver, by default containerd/cri uses cgroups fs so you can:
- don't edit kubelet config
- don't pass `--config`
If you haven't configured containerd to use the systemd driver. Otherwise do the config file above with `cgroupDriver: systemd`.

If kubeadm is wrongly detecting docker for some reason, you can specify `--cri-socket=/run/containerd/containerd.sock` when calling `kubeadm init`

see:
- https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#installing-runtime
- https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/create-cluster-kubeadm/#initializing-your-control-plane-node

**Comment by imoisharma** [comment]
`root@ip-172-31-93-56:/var/lib/kubelet# `kubeadm init --config=kubeadm-init-config.yaml 
no InitConfiguration or ClusterConfiguration kind was found in the YAML file
To see the stack trace of this error execute with --v=5 or higher

It's giving me this error, I am not sure why? Also, I didn't provide any field in config.yaml. When I open it, it's like as I mentioned above.

`root@ip-172-31-93-56:/var/lib/kubelet#` cat kubeadm-init-config.yaml 
`apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
cgroupDriver: cgroupfs 
`

**Comment by BenTheElder** [comment]
if you're using cgroupfs you can skip the `--config` entirely as mentioned above, but you probably want to pass `--cri-socket` (see above), sounds like `kubeadm` is not auto-detecting the socket.

**Comment by imoisharma** [comment]
I tried but it's giving me an error. What am I doing wrong now?
![image](https://user-images.githubusercontent.com/30807347/107846345-2486c200-6e37-11eb-99cf-698ebce7b81c.png)

**Comment by BenTheElder** [comment]
The correct command is: `kubeadm init --cri-socket=/run/containerd/containerd.sock` (no `.yaml` argument).

don't pass `kubeadm-init-config.yaml` by itself, if you supply that pass it with `--config=kubeadm-init-config.yaml` so it is a flag + flag value instead of an argument / subcommand. in this case though you shouldn't need a `--config`.

**Comment by imoisharma** [comment]
I tried the above-mentioned command but still facing the same issue of not recognizing the path.
![image](https://user-images.githubusercontent.com/30807347/107846502-564c5880-6e38-11eb-90f5-0f1e28ed0499.png)

**Comment by BenTheElder** [comment]
Setting `--cri-socket` should fix this, that looks like a bug, except that it looks like you still have a past attempt running (e.g. port 6443 will be the apiserver from previous cluster setup?) `kubeadm reset` can undo.

cc @neolit123, I thought there was a bug about --cri-socket handling a while back but I'm not seeing it, I only see. https://github.com/kubernetes/kubeadm/issues/2270 which was fixed by 1.20.0

**Comment by imoisharma** [comment]
I reset the kubeadm and run it again with ` kubeadm init --cri-socket=/run/containerd/containerd.sock` Still it's saying the path is not found. (And, I was trying to set up earlier on this server. But ain't go sucess)

![image](https://user-images.githubusercontent.com/30807347/107846939-e17b1d80-6e3b-11eb-9508-3be66205de74.png)

**Comment by neolit123** [comment]
@BenTheElder 

> cc @neolit123, I thought there was a bug about --cri-socket handling a while back but I'm not seeing it, I only see. kubernetes/kubeadm#2270 which was fixed by 1.20.0

probably referring to the bug when mixing flag + config for CRI socket:
https://github.com/kubernetes/kubeadm/issues/2039
this is still open.

@imoisharma 

note that mixing `--config` with other flags is not encouraged.
most options are available via the config file.

> I tried but it's giving me an error. What am I doing wrong now?

looks like passing the socket fixed the error for you as `init` succeeded.

the unnecessary calls to `docker info` for some commands were fixed here.
https://github.com/kubernetes/kubeadm/issues/2364

> Kubernetes version (use kubectl version): 1.20.1

the warning during `token create` should be gone in 1.20.2+.

/close
